### PR TITLE
Initial 🗡🩸 at Numeric Arrays means in 🎲

### PR DESCRIPTION
**Pull request by slobodan-ilic**
This PR uses two new JSON payloads obtained from production. It's generated by using the `simple_alltypes` dataset and adding a new `"numeric array"` variable to it (trough superadmin). It aims to demonstrate how we need to treat measures for variables that include additional dimensions (which changes the original shape of the measure). For example, the `Nub_` measure becomes a `Stripe` when this additional (subvariables) dimension is present. `Stripe_` becomes the matrix.

The work is currently done in a way that expands on the functionality of the measure classes, so that they're aware of their own shape (that might be different than the shape of the original cube, because of the ☝️ reasons).

We could do this more rigorously, and implement a separate measure class for the cases when there are subvariables present or not. But given that the measure classes already handle different shapes (nub, stripe, matrix), I've opted for the simpler solution - just being _aware_ of the shape.

### CODE REVIEW

File: src/cr/cube/cube.py [hunk 1]
```diff
@@ -595,6 +598,21 @@ def __init__(self, cube_dict, all_dimensions):
...
+        When we have variables in measures' specification, that are arrays, or when we
+        specify measures such as correlation (that are themselves matrices), this shape
+        needs to change.
+
+        Each subclass is responsible for implementing the correct shape, based on the
+        shape of the provided variable (and if it has subvariables or not), and on the
+        provided measure, and its innate shape.
+        """
+        return self._all_dimensions.shape
+
```
Comment (ernestoarbitrio) [inline src/cr/cube/cube.py:615]: keep the alphabetical order for class property

File: src/cr/cube/cube.py [hunk 2]
```diff
@@ -620,6 +638,30 @@ def _flat_values(self):  # pragma: no cover
...
+        # Original shape: (m, n, . . .,)
+        shape = super(_MeanMeasure, self).shape
+
+        if self._subvariables is not None:
+            # There are subvariables in the provided variable => augment shape
+            # Shape now becomes: (m, n, . . ., n_subvars,)
+            shape += (len(self._subvariables),)
+
+        return shape
+
```
Comment (ernestoarbitrio) [inline src/cr/cube/cube.py:664]: I'd like to see a unit test for this specific property, maybe parametrised to cover the more meaningful edge cases

File: src/cr/cube/cube.py [hunk 3]
```diff
@@ -263,6 +262,12 @@ def has_means(self):
         """True if cube includes a means measure."""
         return self._measures.means is not None
 
+    @lazyproperty
+    def means_subvariables(self):
+        if not self.has_means:
+            return None
+        return self._measures.means.subvariables
+
```
Comment (ernestoarbitrio) [inline src/cr/cube/cube.py:270]: Also this property should be right before the `missing` one. just to respect the alphabetical oder

File: src/cr/cube/cube.py (hunk 2 above)
Comment (slobodan-ilic) [reply src/cr/cube/cube.py:673]: Done

File: src/cr/cube/cube.py (hunk 3 above)
Comment (ernestoarbitrio) [reply src/cr/cube/cube.py]: This name `means_subvariables` is odd to me 🤷‍♂️  , can you explain me what's a subvariable of means? I mean i got from the code that the subvariables belongs to a variable and they're than retrieved in the MeanMeasure, but i don't get te sense of this name ... sorry

File: tests/unit/test_cube.py [hunk 4]
```diff
@@ -519,3 +519,22 @@ def but_the_fraction_is_NaN_for_unfiltered_count_zero(self):
...
+        _base_measure_shape.return_value = expected_shape
+        measure = _MeanMeasure(None, None)
+        assert measure.shape == expected_shape
+
+    def but_it_augments_measure_shape_when_it_has_subvars(self, _base_measure_shape):
+        base_shape = (3, 4)
+        _base_measure_shape.return_value = base_shape
+        measure = _MeanMeasure(None, None, subvariables=["0001", "0002"])
+        expected_shape = base_shape + (2,)  # augmented by n_subvars
+        assert measure.shape == expected_shape
```
Comment (ernestoarbitrio) [inline tests/unit/test_cube.py:536]: Can you collapse this 2 tests onto 1? I mean using `@pytest.mark.parametrize` ... something like:

```python
    @pytest.mark.parametrize(
        "base_shape, subvariables, expected_shape",
        (((3, 4), None, (3, 4)), ((3, 4), ["0001", "0002"], (3, 4, 2))),
    )
    def it_knows_its_altered_shape(
        self, base_shape, subvariables, expected_shape, _base_measure_shape
    ):
        _base_measure_shape.return_value = base_shape
        measure = _MeanMeasure(None, subvariables=subvariables)
        assert measure.shape == expected_shape
```

File: tests/integration/test_numeric_array.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import numpy as np
+
+from cr.cube.cube import Cube
+
+from ..fixtures import NA
+
+NAN = np.nan
```
Comment (ernestoarbitrio) [inline tests/integration/test_numeric_array.py:13]: where is this COSTANT used? i don't see any usage in this file ... am I missing something?

File: src/cr/cube/cube.py [hunk 6]
```diff
@@ -603,11 +611,26 @@ def raw_cube_array(self):
         response. Specifically, it includes values for missing elements, any
         MR_CAT dimensions, and any prunable rows and columns.
         """
-        array = np.array(self._flat_values).reshape(self._all_dimensions.shape)
+        array = np.array(self._flat_values).reshape(self.shape)
```
Comment (ernestoarbitrio) [inline src/cr/cube/cube.py:614]: What do you think if we change the name `array` in something more specific? I know it's not related to you changes but considering that you touch this line, maybe this can be refactored using a meaningful name.

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
Couple of comments + coverage must remain the same (100%) so maybe there's something to be covered in our test suite 😄

File: tests/integration/test_numeric_array.py (hunk 5 above)
Comment (slobodan-ilic) [reply tests/integration/test_numeric_array.py]: Good catch! It was used in the middle of the process but I ended up removing it. Fixed now.

File: tests/unit/test_cube.py (hunk 4 above)
Comment (slobodan-ilic) [reply tests/unit/test_cube.py]: done

File: src/cr/cube/cube.py (hunk 6 above)
Comment (slobodan-ilic) [reply src/cr/cube/cube.py]: Done, I've used the `raw_cube_array` (same as the name of the property) instead. I'm open to better ideas.

File: tests/fixtures/numeric_arrays/means-grouped-by-single-categorical.json [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+                                "?": -8
+                            },
+                            {
+                                "?": -8
+                            }
+                        ]
+                    ],
+                    "metadata": {
+                        "derived": true,
+                        "references": {},
```
Comment (malecki) [inline tests/fixtures/numeric_arrays/means-grouped-by-single-categorical.json:191]: Hey, we *desperately* need the references to be passed along here. In the “movie ratings” example — we know the ratings, but now we have no idea what the movies were! Or even the variable name to know that it’s movies and not magazines! Argh!

Comment (slobodan-ilic) [reply tests/fixtures/numeric_arrays/means-grouped-by-single-categorical.json:191]: We have the `"subvariables"` a couple of lines below. Is that not enough?

File: tests/fixtures/numeric_arrays/means-grouped-by-single-categorical.json [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+                        "type": {
+                            "class": "numeric",
+                            "integer": true,
+                            "missing_reasons": {
+                                "NaN": -8,
+                                "No Data": -1
+                            },
+                            "missing_rules": {},
+                            "subvariables": [
```
Comment (slobodan-ilic) [inline tests/fixtures/numeric_arrays/means-grouped-by-single-categorical.json:200]: @malecki here

File: src/cr/cube/cube.py [hunk 9]
```diff
@@ -312,6 +311,12 @@ def is_weighted(self):
         """True if cube response contains weighted data."""
         return self._measures.is_weighted
 
+    @lazyproperty
+    def means_subvariables(self):
+        if not self.has_means:
```
Comment (ernestoarbitrio) [inline src/cr/cube/cube.py:316]: Considering that we're trying to keep cr.cube well documented I'd like to see a docstring here :D

File: src/cr/cube/cube.py [hunk 10]
```diff
@@ -620,6 +643,34 @@ def _flat_values(self):  # pragma: no cover
 class _MeanMeasure(_BaseMeasure):
     """Statistical mean values from a cube-response."""
 
+    def __init__(self, cube_dict, all_dimensions, subvariables=None):
+        super(_MeanMeasure, self).__init__(cube_dict, all_dimensions)
+        self._subvariables = subvariables
+
+    @lazyproperty
+    def subvariables(self):
+        return self._subvariables
```
Comment (ernestoarbitrio) [inline src/cr/cube/cube.py:652]: docstring also here.

File: src/cr/cube/stripe.py [hunk 11]
```diff
@@ -581,6 +584,8 @@ def count(self):
 
     @lazyproperty
     def mean(self):
+        if self._subvariables is not None:
+            return tuple(self._mean)
         return self._mean
```
Comment (ernestoarbitrio) [inline src/cr/cube/stripe.py:589]: docstrin here

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
some further minor comments

File: src/cr/cube/cube.py (hunk 9 above)
Comment (slobodan-ilic) [reply src/cr/cube/cube.py:317]: done

File: src/cr/cube/cube.py (hunk 10 above)
Comment (slobodan-ilic) [reply src/cr/cube/cube.py]: done

File: src/cr/cube/stripe.py (hunk 11 above)
Comment (slobodan-ilic) [reply src/cr/cube/stripe.py]: done

**Comment by slobodan-ilic** [comment]
Closing in favor of #238
